functions/src/notifications.ts

```diff
@@ -92,6 +92,75 @@ export function createActivityNotification(token: string, activity_id: string):
     }
 }
 
+// Sends a message to a chat if a user leaves or joins it:
+export async function onActivityUsersChanged(change: Change<FirebaseFirestore.DocumentSnapshot>, context: EventContext) {
+    const activityId: string = change.after.id
+    const usersBefore = getUsersFromChange(change.before)
+    const usersAfter = getUsersFromChange(change.after)
+
+    if (usersBefore.length !== usersAfter.length) {
+        // Gets which user joined or left the activity. changedUser will be that person's UID.
+        const changedUser = getDifferentString(usersBefore, usersAfter)
+        if (! changedUser) return -1
+
+        // @ts-ignore because it doesn't like the mock
+        const userInfo = await Refs(database).user(changedUser).get()
+        if (userInfo.exists) {
+            const changedUserInfo = userInfo.data()
+            // Setup message based on whether they left or joined:
+            const msgBody = `${changedUserInfo.name} has ${usersBefore > usersAfter ? 'left' : 'joined'} your activity.`
+            
+            return exports.sendChatMessage(activityId, msgBody, changedUser, new Date())
+        }
+    }
+    return 0
+}
+
+// Sends a user left/joined message to the specified activity
+export function sendChatMessage(activityId: string, message: string, sender: string, date: Date) {
+    // @ts-ignore because it doesn't like the database mock:
+    const chatRef = Refs(database).chat(activityId)
+    return chatRef.add({
+        message,
+        sender,
+        date_sent: date,
+        type: 'user_join_or_leave'
+    })
+}
+
+// Gets the the array of activity members from the activity doc.
+export function getUsersFromChange(doc: FirebaseFirestore.DocumentSnapshot): string[] {
+    const data1 = doc.data()
+    const arr1: string[] = (data1 && data1.members && Array.isArray(data1.members)) ? data1.members : []
+    return arr1
+}
+
+// Returns the first string which is different in the array.
+// Returns null if the arrays are the same length.
+export function getDifferentString(before: string[], after: string[]): string | null {
```

Comment:
Would we ever run into a problem when two+ leave nearly simultaneously?

Here's what I mean:

The event will be triggered twice almost simultaneously.

By the time you get the snapshot, `old = [a, b, c] and new = [c]`

We report `a` as leaving, but then the second trigger also has `old = [a, b, c] and new = [c]` and _also_ reports `a` as leaving.

Reply:
A related question:  Will there ever be _two_ users removed at once, and this function doesn't report return the second?

Reply:
No - I don't believe so. The change should be atomic (right word?) for each change. I think this situation would just trigger twice, once for `old = [a, b, c], new = [b, c]` and then again with `old = [b, c], new = [c]`.

For the other question, we _could_ code it like that but we wouldn't have to.

If we use difference with lodash, both of these situations would be mitigated.

Reply:
Awesome! We just would nee to make sure the message is sent for however many users were in the difference (if we decide to protect against this case, which as you say might very well be impossible)